Added Docker support

patricklecuyer opened the pull request:
Made modifs to ease dockerized deployment.  List of changes:

* Modified scripts to allow taking config options from env variables rather than hardcoded
* Added Dockerfile to create docker image
* Create entrypoint script to initialize database and start bot or frontend
* Updated documentation.

patricklecuyer committed:
Added Docker support

patricklecuyer commented:
Travis CI build fails because of a pre-existing bug on master.  Added PR #10 to fix that bug.

eastebry commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+env PYTHONPATH $PYTHONPATH:/securitybot
+
+RUN apt-get update && \
+    apt-get install -y mysql-client && \
```
Perhaps rather than installing mysql in this container we could link to a separate mysql container, and include a `docker-compose file.`

patricklecuyer replied:
I'm not installing MySQL server in the container.  Just installing client and libs as it's a prereq for the python MySQLDB lib.  I'm also using the client to test for database connection and presence before starting the bot.  I could probably change these test to a Python script rather than call the client directly, should seemed to be less trouble this way.

For deployment, i'm working on a docker-compose with MySQL and a Kubernetes deployment manifest in a separate repo.  Can add them here when they're ready if you'd like,

eastebry replied:
Ah, I see. I think a docker-compose file with a MySQL linked container would be useful in this repository - makes it quite simple to get everything un and running locally. 

The kubernetes config files seem a little out of scope, and something that probably wouldn't be maintained here. Let's leave that in another repository :)

patricklecuyer replied:
Done, added docker-compose with DB, bot and frontend.  Agreed kubernetes deployment is out of scope here, will keep it separate.

eastebry commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+
+WORKDIR /securitybot
+
+ENTRYPOINT ["/securitybot/docker_entrypoint.sh"]
```
Per docker best practices, we should avoid running securitybot as root in the container. How about adding a new user, then adding: `USER securitybot`

patricklecuyer replied:
Agreed, fixed.

patricklecuyer replied:
That was dumb of me, didn't create the user properly, and hadn't tested rebuild properly.  Fixed now and added docker test to travis.

eastebry commented on README.md:
```diff
@@ -33,6 +33,42 @@ If the following were all generated successfully, Securitybot should be up and r
+Dockerfile is included to generate a Docker Image to run the bot.  Entrypoint script will wait for database startup and initialize database if it does not already exist.  Entrypoint takes one of two arguments:
+
+* bot: Starts the main bot
+* fronend: starts the frontend and API server
```
Spelling: frontend

eastebry commented on README.md:
````diff
@@ -33,6 +33,42 @@ If the following were all generated successfully, Securitybot should be up and r
+
+Bot:
+```
+docker run -e DB_NAME=securitybot -e DB_USER=root -e DB_HOST=127.0.0.1 -e DB_PASS=password -e SLACK_API_TOKEN=<your api token> -e DUO_INTEGRATION_KEY=<your integration key> -e DUO_SECRET_KEY=<your secret key> -e DUO_ENDPOINT=<your endpoint> quay.io/patl/securitybot bot
````
We'd prefer to host this image under the dropbox user on the docker registry, but aren't quite ready for that yet.

For now, how about we add instructions to build then run the container. i.e:
```
docker build --tag securitybot 
docker run -e DB_NAME=securitybot -e DB_USER=root -e DB_HOST=127.0.0.1 -e DB_PASS=password -e SLACK_API_TOKEN=<your api token> -e DUO_INTEGRATION_KEY=<your integration key> -e DUO_SECRET_KEY=<your secret key> -e DUO_ENDPOINT=<your endpoint> securitybot bot`
```

Alternatively, the docker-compose file will make running locally a much simpler process.

patricklecuyer replied:
Agreed.  Changed to add description for build.  Will see to add a docker-compose from my other repo later.

eastebry reviewed (changes requested):
Thanks for the PR! Just a few comments

Commit:
Fixed Dockerfile and documentation

eastebry commented on Dockerfile:
```diff
@@ -2,6 +2,8 @@ FROM python:2.7
 
 COPY . /securitybot
 
+USER securitybot
```
I believe this is going to throw an error, since the user will not exist.

Adding a new user up above is the way to go, something like this:
`RUN adduser --disabled-password --gecos '' securitybot`

patricklecuyer committed:
Fixed Dockerfile and added docker build test to Travis

patricklecuyer committed:
Added Docker tests to travis

patricklecuyer committed:
Added docker-compose

patricklecuyer committed:
Fixed travis.yml for Docker tests

Fixed business hours issue for test_auto_escalate

Workaround for flake8 F401 error on type hints

fixed .travis.yml for Docker tests

patricklecuyer committed:
Merge branch 'master' into patl-dockerization

eastebry commented:
Looks good to me. Luke's going to give it one last look through before merging. Thanks for the PR!

patricklecuyer committed:
Added volume to docker-compose

patricklecuyer committed:
Typo

patricklecuyer committed:
Fixed env vars in doc

patricklecuyer commented:
Made some minor changes (mostly cosmetic in docs, had an env var defined twice, and one missing).  Also fixed docker-compose file to add a persistent volume for db.  Let me know if you need anything else for a merge.

NeckBeardPrince commented:
Love to get this in

lfaraone reviewed (approved):
LGTM, assuming tested.
